### ISSUE TITLE
[BUG] AnimatePresence not working propery with Next.js routing

### ISSUE THREAD

**Issue by ronaldruzicka** [framer/motion#474]
**2. Describe the bug**

When I'm trying to animate between different pages the `exit` animations don't work properly, only the `enter` animation works. This problem only occurs in Next.js project. When using classic React Router the animations work normally. Moreover if I go from page `/single/1` to `/single/1` it doesn't animate at all. When changing the base `url` e.g. `/gallery` to `/single/1` only the `enter` animations work.

**3. IMPORTANT: Provide a CodeSandbox reproduction of the bug**

React Router version that works
https://codesandbox.io/s/framer-motion-x-react-router-5l6xd

Next.js version that doesn't work
https://codesandbox.io/s/cold-resonance-tyzh4

You can see that the animations are different in Next.js version, and if you click on **Next** button, it doesn't animate at all.

**4. Steps to reproduce**

1. Go to root
2. Click on image
3. It should scale down and transition to image detail, but the scale down animation doesn't work as in React Router version
4. Click on **Next** button, it doesn't animate at all

**5. Expected behavior**

It should animate also on `exit` and also when changing just `index`/`id` in the route

**7. Environment details**

MacOS Catalina, Chrome 80

Not sure if it's because of some Next.js optimisations or because of SSR. Any help is appreciated 🙏🏻

**Comment by imjoshellis** [comment]
In your next.js app.js, you have a typo. `render()` is `ender()` which is doing some weird things for you. 

If you fix `ender()` to `render()`, you'll get an error because you're using `useRouter()` in a class, when it should be used only in functional components (since it's a hook). 

You should either use `withRouter()` if you want to keep it as a class (see Nextjs Docs) or refactor to a functional component. 

In this case, I forked your Nextjs setup and did a quick-and-dirty refactor to functional component and the animations work as expected: https://codesandbox.io/s/vigorous-dew-l7hig

**Comment by ronaldruzicka** [comment]
Thanks! Yeah, sorry about that, I was copy-pasting and didn't realise it. I fixed it and it's working. But on my project it still doesn't, maybe I'm also missing something. I'll try to create a sandbox.

**Comment by imjoshellis** [comment]
True -- I didn't catch that issue. 

That's because you're dynamically loading the image `src` based on the route id, and the route is set to `/` when you click on "back" (aka before/during the exit animation), meaning the `src` becomes `https://static1.squarespace.com/static/5b475b2c50a54f54f9b4e1dc/t/undefined.jpg?format=1500w`. 

You'd have to change something about the way you're using state in order to make that not happen. I updated [id].js with the `useState()` hook to quickly show one way of accomplishing this, though I'm not sure if it's necessarily the *best* way:

```
const Single = () => {
  const router = useRouter();
  const [imgID, setImgID] = React.useState(
    images[parseInt(router.query.id, 10)]
  );
  return (
    <motion.div className="single" initial="exit" animate="enter" exit="exit">
      <motion.img
        variants={imageVariants}
        src={`https://static1.squarespace.com/static/5b475b2c50a54f54f9b4e1dc/t/${imgID}.jpg?format=1500w`}
        alt="The Barbican"
      />
      <motion.div className="back" variants={backVariants}>
        <Link href="/">
          <a>← Back</a>
        </Link>
      </motion.div>
      <motion.div className="next" variants={backVariants}>
        <Link as={`/image/${Number(router.query.id) + 1}`} href="/image/[id]">
          <a>Next -></a>
        </Link>
      </motion.div>
    </motion.div>
  );
};
```

Sandbox: https://codesandbox.io/s/vigorous-dew-l7hig

**Comment by ronaldruzicka** [comment]
Please could you look at this codesandbox: https://codesandbox.io/s/next-js-framer-motion-vf1k1?

It's pretty much the same structure that I have on my project. If you click on the title (STORY or NEWS) you can see that it goes to different route and therefore animation works. But if you click on those small lines at the bottom of the page, it just changes the index and the animation doesn't work properly. It jumps to next page index, the animation fades out the already new text and then fades in the same. But it should firstly fade out the old and fade in the new.

Maybe I'm again missing something small, but I cannot figure it out.

**Comment by imjoshellis** [comment]
That sandbox doesn't appear to be rendering correctly on my end. Can you double-check it? I'm just getting a white screen in the sandbox's browser. Tried viewing in a few different browsers.

**Comment by ronaldruzicka** [comment]
Sorry the root path wasn't set properly. I fixed it now. Just click the `button` Story and it will go to News and than News will go back to Story. Or just go directly here: https://vf1k1.sse.codesandbox.io/story/0

**Comment by imjoshellis** [comment]
I tried a few different ideas, but the only one I got to work properly is `useState` (again): https://codesandbox.io/s/next-js-framer-motion-oo2dd

The only line I changed was this on both `[index].js` files: ` const [currentPageIndex] = useState(Number(router.query.index) || 0);`

Without using state, you're changing the content of `<PageDetailLayout>` (ie `<Page>`) instantly when the route changes. Framer-motion was correctly holding the original component and animating it out, but what you see is this:

1. Page 1 content changes to the content of page 2 because the route changed (framer doesn't deal with content changes, just component mount/unmount)
2. Page 1 animates out
3. Page 2 animates in

Using state allows for this:

1. Page 1 content doesn't change even though the route changed
2. Page 1 animates out
3. Page 2 animates in
